### ISSUE TITLE
Mouse events are ignored after cursor leaves window

### ISSUE THREAD

**Issue by jakobhellermann** [mvlabat/bevy_egui#61]
Before https://github.com/mvlabat/bevy_egui/commit/a496e4984b7ab9224d2a33571187fd188da03add it was possible to start dragging somewhere in the window (for example a slider), then move the cursor out of the window while still having the mouse button pressed and then returning.

After the change, the mouse events will stop firing as soon as the cursor leaves the window.

Removing the `if` in

https://github.com/mvlabat/bevy_egui/commit/a496e4984b7ab9224d2a33571187fd188da03add#diff-420f4e08a25299e0e1fb09f4757e1d5d027a2278f37030c09ab9c06786cfa52eR155-R157

fixes this, although I'm not sure why, because the `if` condition is only `false` for one frame.

Before:

https://user-images.githubusercontent.com/22177966/150784489-6eb90928-5989-45bf-a7f9-f7e13efaa064.mp4

After:

https://user-images.githubusercontent.com/22177966/150785233-1d3dc145-e37e-424f-a990-8d8055570ef5.mp4

**Comment by mvlabat** [comment]
I believe it's by Egui's design that when you emit `egui::Event::PointerGone` and don't push another `egui::Event::PointerMoved` event during the same frame, it forgets that the mouse button is still being pressed and the cursor started its movement from a certain widget. I can reproduce the same behaviour in the [Egui's demo](https://emilk.github.io/egui/). So I'd even consider the previous behaviour to be a bug, as pushing `egui::Event::PointerMoved` right after `egui::Event::PointerGone` doesn't quite make sense.

Though I agree that being able to keep dragging a slider outside a window is a good thing. But I guess it's an issue of Egui itself.

**Comment by coderedart** [comment]
I am kinda stuck on this. 
1.  so, with eframe, if i click and drag beyond the bounds of native window, and release the mouse button, egui will stop the drag. it works as expected.
2. but, with bevy_egui, if i click and drag beyond the bounds of native window, even after releasing the mouse button, egui will still continue the drag, which leads to egui not using the events at all. so, I have to click somewhere again for it to stop the drag.

https://user-images.githubusercontent.com/24411704/166121391-cf3791c9-e472-463f-aa41-66372a5a7a85.mp4

the above video just has a `PointerState` ui and a list of latest events (except the cursor moved event to avoid clutter). as you can see, we get `PointerGone` event twice, but we do not get the PointerButton released event at all. so, until that button is released by clicking again somewhere, we cannot start interacting with egui again :( I checked the event reader and there is definitely a MouseButtonInput event with the release.


bevy_egui doesn't collect MouseButton events if the cursor left that window it seems :( https://github.com/mvlabat/bevy_egui/blob/main/src/systems.rs#L171 . but some window backends like glfw as well as winit send a fake key/button released event when the window loses focus or loses cursor, which come after the Cursor Left event. 

Is there any reasoning as to why skip MouseButton events if Cursor left the window?
in the same spirit, is there a reason why only the focused window get the `Key` events, and the rest ignored? I would assume the same issues like Key Release event sent after the window losing focus can happen. 

for example, press `A`, click on another window /desktop to lose focus and winit sends a release event, but bevy_egui still keeps the A pressed as true even after i focus the egui window while pressing other buttons, `A` will remain in `key_down` status until I press `A` again.

**Comment by coderedart** [comment]
after digging in a little deeper, it seems bevy sends MouseButtonInput and Key events to a global Events resource. but, CursorEnter or WindowFocus kind of events to per window (with the window ID mentioned inside that event). 

I sort of understand why things ended up this way in bevy_egui.
lets say, we have two windows. window A is in focus. if player clicked on window B, then we have Cursor Left event for window A, CursorEnter event for window B and ofcourse a MouseButton Press/Release in global reader. we have no way of knowing whether the MouseButton events belong to window A  before its CursorLeft event or window B after the CursorEnter event. :( 

We can do this three ways:
1. request egui library to clear all the "pressed" keys / mouse_buttons when it receives a PointerGone event. solves part of the problem, but we still have the issue of older events before CursorEnter of window B and it is not possible to reliably know to which window a MouseButtonInput or Keyboard event belongs to.
2. request bevy to allow specifying window ID for MouseButton and Keyboard Input. this is the only clean way to handle this in the long term. 
3. hacky workarounds like ignoring the pressed events when there's a cursor enter / left event, and sending release events to all windows who had a cursor enter / left event happen on them for two frames. this will miss the first click on window B, but we don't have to ask egui or bevy to change anything. but ofcourse, maintenance burden, especially considering other plugins like bevy_egui_inspect depending on this crate.

**Comment by mvlabat** [comment]
The raised issue was fixed in the a3d113ae099626a4a3999a6fc15b05d726d99c0c commit, but caused a new one: #101. I believe it's a better compromise for now, but we definitely should look into that one as well.

@coderedart thanks a lot for your thorough research. For some reason, the behaviour I observed was different from the one you show in the video, but it actually started behaving like that in browsers once I made the mentioned commit. For some reason, Bevy fails to report mouse up events if a cursor is outside a window at all.

I'm closing this issue, so I suggest continuing the discussion in the new one.
